### PR TITLE
test: standardize ControlPlaneWaiters

### PR DESCRIPTION
**Pull request by jackfrancis**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

This PR creates a single, standard `EnsureControlPlaneInitialized` helper to accommodate all scenarios. Within that standard helper we can introspect the cluster configuration to determine whether or not various helm charts need to be employed to help bootstap the cluster (e.g., cloud-provider-azure). This has the outcome of de-coupling particular helm foo from particular cluster scenarios, making future E2E maintenance more easy and fluid.

We also clarify some of the AKS E2E helpers so they more accurately express that they are designed for AKS scenarios only.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/aks.go [hunk 1]
```diff
@@ -39,41 +39,41 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// WaitForControlPlaneInitialized waits for the azure managed control plane to be initialized.
+// DiscoverAndWaitForAKSControlPlaneInput contains the fields the required for checking the status of azure managed control plane.
```
Comment (mboersma) [inline test/e2e/aks.go:42]:
```suggestion
// DiscoverAndWaitForAKSControlPlaneInput contains the fields required for checking the status of an Azure managed control plane.
```

File: test/e2e/aks.go [hunk 2]
```diff
@@ -39,41 +39,41 @@ import (
...
+type DiscoverAndWaitForAKSControlPlaneInput struct {
+	Lister  framework.Lister
+	Getter  framework.Getter
+	Cluster *clusterv1.Cluster
+}
+
+// WaitForAKSControlPlaneInitialized waits for the azure managed control plane to be initialized.
```
Comment (mboersma) [inline test/e2e/aks.go:49]:
```suggestion
// WaitForAKSControlPlaneInitialized waits for the Azure managed control plane to be initialized.
```

File: test/e2e/aks.go [hunk 3]
```diff
@@ -39,41 +39,41 @@ import (
...
-	Cluster *clusterv1.Cluster
-}
-
-// DiscoverAndWaitForControlPlaneInitialized gets the azure managed control plane associated with the cluster,
-// and waits for at least one control plane machine to be up.
-func DiscoverAndWaitForControlPlaneInitialized(ctx context.Context, input DiscoverAndWaitForControlPlaneMachinesInput, intervals ...interface{}) {
-	Expect(ctx).NotTo(BeNil(), "ctx is required for DiscoverAndWaitForControlPlaneInitialized")
-	Expect(input.Lister).NotTo(BeNil(), "Invalid argument. input.Lister can't be nil when calling DiscoverAndWaitForControlPlaneInitialized")
-	Expect(input.Cluster).NotTo(BeNil(), "Invalid argument. input.Cluster can't be nil when calling DiscoverAndWaitForControlPlaneInitialized")
+// DiscoverAndWaitForAKSControlPlaneInitialized gets the azure managed control plane associated with the cluster,
```
Comment (mboersma) [inline test/e2e/aks.go:71]:
```suggestion
// DiscoverAndWaitForAKSControlPlaneInitialized gets the Azure managed control plane associated with the cluster
```

File: test/e2e/aks.go [hunk 4]
```diff
@@ -92,12 +92,12 @@ func DiscoverAndWaitForControlPlaneInitialized(ctx context.Context, input Discov
 	}, intervals...)
 }
 
-// DiscoverAndWaitForControlPlaneReady gets the azure managed control plane associated with the cluster,
-// and waits for all the control plane machines to be up.
-func DiscoverAndWaitForControlPlaneReady(ctx context.Context, input DiscoverAndWaitForControlPlaneMachinesInput, intervals ...interface{}) {
-	Expect(ctx).NotTo(BeNil(), "ctx is required for DiscoverAndWaitForControlPlaneReady")
-	Expect(input.Lister).NotTo(BeNil(), "Invalid argument. input.Lister can't be nil when calling DiscoverAndWaitForControlPlaneReady")
-	Expect(input.Cluster).NotTo(BeNil(), "Invalid argument. input.Cluster can't be nil when calling DiscoverAndWaitForControlPlaneReady")
+// DiscoverAndWaitForAKSControlPlaneReady gets the azure managed control plane associated with the cluster,
```
Comment (mboersma) [inline test/e2e/aks.go:95]:
```suggestion
// DiscoverAndWaitForAKSControlPlaneReady gets the Azure managed control plane associated with the cluster
```

File: test/e2e/aks.go [hunk 5]
```diff
@@ -145,16 +145,16 @@ type WaitForControlPlaneAndMachinesReadyInput struct {
 	Namespace    string
 }
 
-// WaitForAtLeastOneControlPlaneAndMachineToExist waits for at least one control plane machine to be provisioned.
-func WaitForAtLeastOneControlPlaneAndMachineToExist(ctx context.Context, input WaitForControlPlaneAndMachinesReadyInput, intervals ...interface{}) {
-	By("Waiting for at least one control plane node to exist")
-	WaitForControlPlaneMachinesToExist(ctx, input, atLeastOne, intervals...)
+// WaitForAtLeastOneSystemNodePoolMachineToExist waits for at least machine in the 'system' node pool to exist.
```
Comment (mboersma) [inline test/e2e/aks.go:148]:
```suggestion
// WaitForAtLeastOneSystemNodePoolMachineToExist waits for at least one machine in the "system" node pool to exist.
```

File: test/e2e/aks.go [hunk 6]
```diff
@@ -145,16 +145,16 @@ type WaitForControlPlaneAndMachinesReadyInput struct {
...
+func WaitForAtLeastOneSystemNodePoolMachineToExist(ctx context.Context, input WaitForControlPlaneAndMachinesReadyInput, intervals ...interface{}) {
+	By("Waiting for at least one node to exist in the 'system' node pool")
+	WaitForAKSSystemNodePoolMachinesToExist(ctx, input, atLeastOne, intervals...)
 }
 
-// WaitForAllControlPlaneAndMachinesToExist waits for all control plane machines to be provisioned.
+// WaitForAllControlPlaneAndMachinesToExist waits for all machines in the 'system' node pool to exist.
```
Comment (mboersma) [inline test/e2e/aks.go:154]:
```suggestion
// WaitForAllControlPlaneAndMachinesToExist waits for all machines in the "system" node pool to exist.
```

File: test/e2e/aks.go [hunk 7]
```diff
@@ -176,8 +176,8 @@ func (r controlPlaneReplicas) value(mp *clusterv1exp.MachinePool) int {
 	return 0
 }
 
-// WaitForControlPlaneMachinesToExist waits for a certain number of control plane machines to be provisioned represented.
-func WaitForControlPlaneMachinesToExist(ctx context.Context, input WaitForControlPlaneAndMachinesReadyInput, minReplicas controlPlaneReplicas, intervals ...interface{}) {
+// WaitForAKSSystemNodePoolMachinesToExist waits for a certain number of machines in the 'system' node pool to exist.
```
Comment (mboersma) [inline test/e2e/aks.go:179]:
```suggestion
// WaitForAKSSystemNodePoolMachinesToExist waits for a certain number of machines in the "system" node pool to exist.
```

File: test/e2e/cloud-provider-azure.go [hunk 8]
```diff
@@ -24,49 +24,30 @@ import (
...
+	cloudProviderAzureHelmReleaseName = "cloud-provider-azure-oot"
 )
 
 // InstallCloudProviderAzureHelmChart installs the official cloud-provider-azure helm chart
-// Fulfills the clusterctl.Waiter type so that it can be used as ApplyClusterTemplateAndWaitInput data
-// in the flow of a clusterctl.ApplyClusterTemplateAndWait E2E test scenario
-func InstallCloudProviderAzureHelmChart(ctx context.Context, input clusterctl.ApplyClusterTemplateAndWaitInput, result *clusterctl.ApplyClusterTemplateAndWaitResult) {
-	By(fmt.Sprintf("Ensuring the kubeconfig secret for cluster %s/%s exists before installing cloud-provider-azure components", input.ConfigCluster.Namespace, input.ConfigCluster.ClusterName))
-	WaitForWorkloadClusterKubeconfigSecret(ctx, input)
+// and validates that expected pods exist and are Ready
```
Comment (mboersma) [inline test/e2e/cloud-provider-azure.go:37]:
```suggestion
// and validates that expected pods exist and are Ready.
```

Comment (mboersma) [reply test/e2e/cloud-provider-azure.go:37]: Obviously not the end of the world, but I thought we had a linter in place that enforced "comments should end with a period." I'll investigate.

File: test/e2e/common.go [hunk 9]
```diff
@@ -215,3 +217,36 @@ func createRestConfig(ctx context.Context, tmpdir, namespace, clusterName string
 
 	return config
 }
+
+// EnsureControlPlaneInitialized waits for the cluster KubeadmControlPlane object to be initialized
+// and then installs cloud-provider-azure components via helm
```
Comment (mboersma) [inline test/e2e/common.go:222]:
```suggestion
// and then installs cloud-provider-azure components via Helm.
```

File: test/e2e/common.go [hunk 10]
```diff
@@ -215,3 +217,36 @@ func createRestConfig(ctx context.Context, tmpdir, namespace, clusterName string
...
+// Fulfills the clusterctl.Waiter type so that it can be used as ApplyClusterTemplateAndWaitInput data
+// in the flow of a clusterctl.ApplyClusterTemplateAndWait E2E test scenario
```
Comment (mboersma) [inline test/e2e/common.go:224]:
```suggestion
// in the flow of a clusterctl.ApplyClusterTemplateAndWait E2E test scenario.
```

File: test/e2e/helpers.go [hunk 11]
```diff
@@ -714,8 +714,11 @@ func resolveKubetestRepoListPath(version string, path string) (string, error) {
 // resolveKubernetesVersions looks at Kubernetes versions set as variables in the e2e config and sets them to a valid k8s version
 // that has an existing capi offer image available. For example, if the version is "stable-1.22", the function will set it to the latest 1.22 version that has a published reference image.
 func resolveKubernetesVersions(config *clusterctl.E2EConfig) {
-	ubuntuVersions := getVersionsInOffer(context.TODO(), os.Getenv(AzureLocation), capiImagePublisher, capiOfferName)
-	windowsVersions := getVersionsInOffer(context.TODO(), os.Getenv(AzureLocation), capiImagePublisher, capiWindowsOfferName)
+	ubuntuSkus := getImageSkusInOffer(context.TODO(), os.Getenv(AzureLocation), capiImagePublisher, capiOfferName)
```
Comment (mboersma) [inline test/e2e/helpers.go:717]: This image helper refactoring seems unrelated to the ControlPlane waiters changes. Could it be a separate PR?

File: test/e2e/helpers.go [hunk 12]
```diff
@@ -909,19 +907,16 @@ func WaitForDaemonset(ctx context.Context, input clusterctl.ApplyClusterTemplate
...
-// podListHasNumPods fulfills the cluster-api PodListCondition type spec
-// given a list of pods, we validate for an exact number of those pods in a Running state
-func podListHasNumPods(numPods int) func(pl *corev1.PodList) error {
-	return func(pl *corev1.PodList) error {
-		var runningPods int
-		for _, p := range pl.Items {
-			if p.Status.Phase == corev1.PodRunning {
-				runningPods++
-			}
+// WaitForDeployment retries during E2E until a deployments's pods are all Running
```
Comment (mboersma) [inline test/e2e/helpers.go:910]:
```suggestion
// WaitForDeployment retries during E2E until a deployments's pods are all Running.
```

File: test/e2e/cloud-provider-azure.go [hunk 13]
```diff
@@ -40,75 +36,42 @@ const (
...
-	WaitForWorkloadClusterKubeconfigSecret(ctx, input)
+// and validates that expected pods exist and are Ready.
+func InstallCloudProviderAzureHelmChart(ctx context.Context, input clusterctl.ApplyClusterTemplateAndWaitInput) {
 	By("Installing the correct version of cloud-provider-azure components via helm")
 	values := []string{fmt.Sprintf("infra.clusterName=%s", input.ConfigCluster.ClusterName)}
 	InstallHelmChart(ctx, input, cloudProviderAzureHelmRepoURL, cloudProviderAzureChartName, cloudProviderAzureHelmReleaseName, values)
 	By("Waiting for a Running cloud-controller-manager pod")
 	clusterProxy := input.ClusterProxy.GetWorkloadCluster(ctx, input.ConfigCluster.Namespace, input.ConfigCluster.ClusterName)
 	workloadClusterClient := clusterProxy.GetClient()
-	cloudControllerManagerPodLabel, err := labels.Parse("component=cloud-controller-manager")
```
Comment (jackfrancis) [inline test/e2e/cloud-provider-azure.go:60]: This manual pod filter and list is too cute by half, and in addition it is less accurate in terms of pod count than relying upon the parent Deployment resource metadata (see the new `WaitForDeployment` helper func).

File: test/e2e/common.go [hunk 14]
```diff
@@ -215,3 +217,39 @@ func createRestConfig(ctx context.Context, tmpdir, namespace, clusterName string
...
+		Namespace: cluster.Spec.ControlPlaneRef.Namespace,
+		Name:      cluster.Spec.ControlPlaneRef.Name,
+	}
+	Eventually(func() error {
+		return getter.Get(ctx, key, kubeadmControlPlane)
+	}, input.WaitForControlPlaneIntervals...).Should(Succeed(), "Failed to get KubeadmControlPlane object %s/%s", cluster.Spec.ControlPlaneRef.Namespace, cluster.Spec.ControlPlaneRef.Name)
+	if kubeadmControlPlane.Spec.KubeadmConfigSpec.ClusterConfiguration.ControllerManager.ExtraArgs["cloud-provider"] == "external" {
+		InstallCloudProviderAzureHelmChart(ctx, input)
+	}
+	if kubeadmControlPlane.Spec.KubeadmConfigSpec.ClusterConfiguration.ControllerManager.ExtraArgs["external-cloud-volume-plugin"] == "azure" {
```
Comment (jackfrancis) [inline test/e2e/common.go:243]: @sonasingh46 to approve this being a definitive configuration trigger for installing the azuredisk CSI helm chart

**Review by mboersma — APPROVED** [review]
/lgtm

File: test/e2e/common.go (hunk 14 above)
Comment (sonasingh46) [reply test/e2e/common.go:243]: This flag is necessary during upgrade to kubernetes version 1.23 that uses external cloud provider otherwise the upgrade will fail. An upgrade to k8s version 1.23 will mean to install external azurediskcsi driver as the azurecsidisk migration is enabled by default. 

So wether external azurediskcsi driver should be installed or not should solely depend on if AzureDiskCSIMigraion is enabled or not -- 
i) For k8s version 1.23 and later it is enabled by default.
ii) For k8s version prior to 1.23, it can be enabled by passing a flag.

TL;DR
At this place, I think it is more helpful if we leave upon the users/callers to decide what helm component they want to install by becoming more explicit. For example, let us say we have 5 helm things -- but as per the specific CI template only 3  installs is required.

Comment (sonasingh46) [reply test/e2e/common.go:243]: One simple way top of my head is something like following: 

```
const (
	AzureDiskCSIDriver = "AzureDiskCSIDriver"
	CloudProviderAzure = "CloudProviderAzure"
)

var Installers = map[string]func(){
	AzureDiskCSIDriver : InstallAzureDiskCSIDriver(),
	CloudProviderAzure : InstallCloudProviderAzure(),
}

func InstallAzureDiskCSIDriver() func(){
	return func() {
		// install code
	}
}

func InstallCloudProviderAzure() func(){
	return func() {
		// install code
	}
}

type HelmInstallOpts struct{
	Packages []string
}

func EnsureControlPlaneInitialized(opts HelmInstallOpts){
	// Already existing code
	for _,p:= range opts.Packages{
		install,ok:=Installers[p]
		if !ok{
			// report error as p is unsupported helm package
		}
		install()
	}
}

```
wdyt? Or any approach that just gives the flexibility to choose what helm to install.

Comment (sonasingh46) [reply test/e2e/common.go:243]: We can do this in a different PR also.

Comment (jackfrancis) [reply test/e2e/common.go:243]: I love that you're thinking of users here, but in this case I think we are focused solely on figuring out how to evaluate from a given cluster template yaml "am I able to install (or do I need to install) this particular addon?" for CI testing only.

I seems to me like the azuredisk-csi drivers actually work in any Kubernetes cluster scenario for the versions that we test, so how about we just install it for every CI test template?

Comment (sonasingh46) [reply test/e2e/common.go]: Okay. So I am coming from a point that -- there are cases where we do not want to install azurediskcsi driver for e2e. 
Like an e2e for workload cluster upgrade from v1.22 to v1.23.

Comment (sonasingh46) [reply test/e2e/common.go]: But the flag check should do fine. Let us go with it and if need arises we will revisit.

**Review by sonasingh46 — APPROVED** [review]

File: test/e2e/helpers.go [hunk 15]
```diff
@@ -911,36 +894,16 @@ func WaitForDaemonset(ctx context.Context, input clusterctl.ApplyClusterTemplate
...
-// given a list of pods, we validate for at least 1 number of those pods in a Running state
-func podListHasAtLeastNumPods(numPods int) func(pl *corev1.PodList) error {
-	return func(pl *corev1.PodList) error {
-		var runningPods int
-		for _, p := range pl.Items {
-			if p.Status.Phase == corev1.PodRunning {
-				runningPods++
-			}
+// WaitForDeployment retries during E2E until a deployments's pods are all Running.
+func WaitForDeployment(ctx context.Context, input clusterctl.ApplyClusterTemplateAndWaitInput, cl client.Client, name, namespace string) {
```
Comment (CecileRobertMichon) [inline test/e2e/helpers.go:898]: we already have 

```
// WaitForDeploymentsAvailable waits until the Deployment has status.Available = True, that signals that
// all the desired replicas are in place.
// This can be used to check if Cluster API controllers installed in the management cluster are working.
func WaitForDeploymentsAvailable(ctx context.Context, input WaitForDeploymentsAvailableInput, intervals ...interface{})
```

could it be used instead of adding this new helper? It essentially does the same thing as far as I can tell

File: test/e2e/cloud-provider-azure.go [hunk 16]
```diff
@@ -40,75 +36,41 @@ const (
...
-		ListOptions: &client.ListOptions{
-			LabelSelector: cloudControllerManagerPodLabel,
-			Namespace:     "kube-system",
-		},
-		Condition: podListHasNumPods(1),
-	}, input.WaitForControlPlaneIntervals...)
-	Expect(err).NotTo(HaveOccurred())
-	By(fmt.Sprintf("Waiting for Ready cloud-node-manager daemonset pods"))
+	By("Waiting for Ready cloud-controller-manager deployment pods")
+	for _, d := range []string{"cloud-controller-manager"} {
```
Comment (CecileRobertMichon) [inline test/e2e/cloud-provider-azure.go:49]: why range over a single string here?

**Review by CecileRobertMichon — APPROVED** [review]
lgtm

**Review by CecileRobertMichon — APPROVED** [review]
/lgtm
/approve

I don't think my string array comment is blocking, it's mostly a style preference
